go/apps/gw/server/middleware_errors.go

```diff
@@ -0,0 +1,168 @@
+package server
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/unkeyed/unkey/go/pkg/codes"
+	"github.com/unkeyed/unkey/go/pkg/fault"
+	"github.com/unkeyed/unkey/go/pkg/otel/logging"
+)
+
+// ErrorResponse is the standard error response format for the gateway.
+type ErrorResponse struct {
+	Meta  Meta        `json:"meta"`
+	Error ErrorDetail `json:"error"`
+}
+
+// Meta contains metadata about the request.
+type Meta struct {
+	RequestID string `json:"requestId"`
+}
+
+// ErrorDetail contains error information.
+type ErrorDetail struct {
+	Title  string `json:"title"`
+	Type   string `json:"type,omitempty"`
+	Detail string `json:"detail"`
+	Status int    `json:"status"`
+}
+
+// WithErrorHandling returns middleware that translates errors into appropriate
+// HTTP responses based on error URNs.
+func WithErrorHandling(logger logging.Logger) Middleware {
+	return func(next HandleFunc) HandleFunc {
+		return func(ctx context.Context, s *Session) error {
+			err := next(ctx, s)
+			if err == nil {
+				return nil
+			}
+
+			// Get the error URN from the error
+			urn, ok := fault.GetCode(err)
+			if !ok {
+				urn = codes.App.Internal.UnexpectedError.URN()
+			}
+
+			code, parseErr := codes.ParseURN(urn)
+			if parseErr != nil {
+				logger.Error("failed to parse error code", "error", parseErr.Error())
+				code = codes.App.Internal.UnexpectedError
+			}
+
+			// Determine status code based on error type
+			status := http.StatusInternalServerError
+
+			switch urn {
+			case codes.UnkeyGatewayErrorsInternalInternalServerError,
+				codes.UnkeyGatewayErrorsInternalKeyVerificationFailed:
+				status = http.StatusInternalServerError
+			case codes.UnkeyGatewayErrorsValidationRequestInvalid,
+				codes.UnkeyGatewayErrorsValidationResponseInvalid:
+				status = http.StatusBadRequest
+			case codes.UnkeyGatewayErrorsAuthUnauthorized:
+				status = http.StatusUnauthorized
+			case codes.UnkeyGatewayErrorsAuthRateLimited:
+				status = http.StatusTooManyRequests
+			// Gateway errors - 404, 502, 503, 504
+			case codes.UnkeyGatewayErrorsRoutingConfigNotFound:
+				status = http.StatusNotFound
+			case codes.UnkeyGatewayErrorsProxyBadGateway:
+				status = http.StatusBadGateway
+			case codes.UnkeyGatewayErrorsProxyServiceUnavailable:
+				status = http.StatusServiceUnavailable
+			case codes.UnkeyGatewayErrorsProxyGatewayTimeout:
+				status = http.StatusGatewayTimeout
+			}
+
+			// Log the error with correct status
+			logger.Error("gateway error",
+				"error", err.Error(),
+				"requestId", s.RequestID(),
+				"publicMessage", fault.UserFacingMessage(err),
+				"status", status,
+			)
+
+			// Handle gateway errors with HTML responses
+			switch urn {
```

Comment:
This produces really ugly errors on the client side if they expected json,

what if we only returned html if they signaled they can handle it via `Accept` header?
Browsers would send this automatically, and sdks could receive the json error

if you agree, let's create an issue for later